File: backend/src/auth/verification.ts
import crypto from 'crypto'
import bcrypt from 'bcryptjs'
import { prisma } from '../db/client.js'
import { findUserByEmail } from '../common/account.js'
import { sendVerificationEmail, sendPasswordResetEmail } from '../notifications/email.js'
import { ErrorCodes, AuthServiceError, faceErrorFromException } from '../common/errors.js'

export function generateToken(): string {
  return crypto.randomBytes(32).toString('hex')
}

export async function issueEmailVerification(userId: string, email: string): Promise<void> {
  const token = generateToken()
  await prisma.user.update({
    where: { id: userId },
    data: { emailVerifyToken: token, emailVerifiedAt: null },
  })
  await sendVerificationEmail(email, token)
}

export async function markEmailVerified(userId: string): Promise<void> {
  await prisma.user.update({
    where: { id: userId },
    data: { emailVerifiedAt: new Date(), emailVerifyToken: null },
  })
}

const lastResendAt = new Map<string, number>()

export async function verifyEmailWithToken(token: string): Promise<void> {
  const user = await prisma.user.findFirst({
    where: { emailVerifyToken: token },
    select: { id: true },
  })
  if (!user) throw new AuthServiceError(400, ErrorCodes.EMAIL_VERIFY_TOKEN_INVALID)
  await markEmailVerified(user.id)
}

export async function verifyEmailAndGetRedirect(token: string): Promise<string> {
  const appUrl = (process.env.APP_PUBLIC_URL ?? 'https://spectrmod.com').replace(/\/$/, '')
  if (!token) return `${appUrl}/verify-email?error=invalid`
  const user = await prisma.user.findFirst({
    where: { emailVerifyToken: token },
    select: { id: true },
  })
  if (!user) return `${appUrl}/verify-email?error=invalid`
  await markEmailVerified(user.id)
  return `${appUrl}/verify-email?verified=1`
}

export async function resendVerification(userId: string): Promise<{ success: true }> {
  const last = lastResendAt.get(userId) ?? 0
  if (Date.now() - last < 60_000) throw new AuthServiceError(429, ErrorCodes.RESEND_COOLDOWN)
  const user = await prisma.user.findUnique({
    where: { id: userId },
    select: { email: true, emailVerifiedAt: true, passwordHash: true },
  })
  if (!user || !user.passwordHash) return { success: true }
  if (user.emailVerifiedAt) return { success: true }
  try {
    await issueEmailVerification(userId, user.email)
    lastResendAt.set(userId, Date.now())
    return { success: true }
  } catch (e) {
    console.error('[resend-verification]', e)
    throw new AuthServiceError(500, ErrorCodes.EMAIL_SEND_FAILED)
  }
}

export async function forgotPassword(email: unknown): Promise<{ success: true }> {
  if (!email || typeof email !== 'string' || !email.includes('@')) {
    throw new AuthServiceError(400, ErrorCodes.INVALID_EMAIL)
  }
  const normalizedEmail = email.toLowerCase().trim()
  const user = await findUserByEmail(normalizedEmail)
  if (!user || user.deletedAt) return { success: true }
  if (!user.passwordHash) throw new AuthServiceError(400, ErrorCodes.OAUTH_ACCOUNT_NO_PASSWORD)

  const token = generateToken()
  const expires = new Date(Date.now() + 3_600_000)
  await prisma.user.update({
    where: { id: user.id },
    data: { passwordResetToken: token, passwordResetExpires: expires },
  })
  try {
    await sendPasswordResetEmail(user.email, token)
  } catch (e) {
    console.error('[forgot-password] send failed:', e)
    throw new AuthServiceError(500, ErrorCodes.EMAIL_SEND_FAILED)
  }
  return { success: true }
}

export async function resetPassword(input: {
  token?: string
  password?: string
}): Promise<{ success: true }> {
  const { token, password } = input
  if (!token || !password) throw new AuthServiceError(400, ErrorCodes.MISSING_FIELD)
  if (password.length < 8) throw new AuthServiceError(400, ErrorCodes.PASSWORD_TOO_SHORT)

  const user = await prisma.user.findFirst({
    where: { passwordResetToken: token, passwordResetExpires: { gt: new Date() } },
  })
  if (!user) throw new AuthServiceError(400, ErrorCodes.PASSWORD_RESET_TOKEN_INVALID)

  await prisma.user.update({
    where: { id: user.id },
    data: {
      passwordHash: await bcrypt.hash(password, 12),
      passwordResetToken: null,
      passwordResetExpires: null,
    },
  })
  return { success: true }
}

import {
  CURRENT_FACE_MODEL,
  embedBurstFromBase64,
  ensureFaceService,
  passesEnrollQuality,
} from '../face/service.js'

const MIN_INPUT_FRAMES = 3
const MAX_INPUT_FRAMES = 16
const MIN_ACCEPTED_EMBEDDINGS = 4

export async function enrollFace(
  userId: string,
  photos: unknown
): Promise<{ success: true; accepted: number; total: number }> {
  if (!photos || !Array.isArray(photos) || photos.length === 0) {
    throw new AuthServiceError(400, ErrorCodes.PHOTOS_REQUIRED)
  }
  if (photos.length < MIN_INPUT_FRAMES || photos.length > MAX_INPUT_FRAMES) {
    throw new AuthServiceError(400, ErrorCodes.FACE_ENROLL_TOO_FEW_FRAMES)
  }
  for (const photo of photos) {
    if (typeof photo !== 'string' || !photo.startsWith('data:image/')) {
      throw new AuthServiceError(400, ErrorCodes.INVALID_PHOTO)
    }
  }

  try {
    await ensureFaceService()
    const results = await embedBurstFromBase64(photos as string[])

    const accepted: {
      vector: number[]
      detScore: number
      yaw: number
      pitch: number
      blurVar: number
    }[] = []
    const reasons: Record<string, number> = {}

    for (const r of results) {
      if (!r.face) {
        const k = r.error ?? 'no_face'
        reasons[k] = (reasons[k] ?? 0) + 1
        continue
      }
      const q = passesEnrollQuality(r.face)
      if (!q.ok) {
        reasons[q.reason ?? 'low_quality'] = (reasons[q.reason ?? 'low_quality'] ?? 0) + 1
        continue
      }
      accepted.push({
        vector: r.face.embedding,
        detScore: r.face.det_score,
        yaw: r.face.yaw,
        pitch: r.face.pitch,
        blurVar: r.face.blur_var,
      })
    }

    console.log(
      `[enroll-face] user=${userId} frames=${photos.length} accepted=${accepted.length} reasons=${JSON.stringify(reasons)}`
    )

    if (accepted.length < MIN_ACCEPTED_EMBEDDINGS) {
      throw new AuthServiceError(400, ErrorCodes.FACE_ENROLL_LOW_QUALITY, undefined, {
        accepted: accepted.length,
        needed: MIN_ACCEPTED_EMBEDDINGS,
        reasons,
      })
    }

    await prisma.$transaction(async (tx) => {
      await tx.faceEmbedding.deleteMany({ where: { userId } })
      await tx.faceEmbedding.createMany({
        data: accepted.map((a) => ({
          userId,
          vector: a.vector,
          detScore: a.detScore,
          yaw: a.yaw,
          pitch: a.pitch,
          blurVar: a.blurVar,
          faceModel: CURRENT_FACE_MODEL,
          source: 'enrollment',
        })),
      })
      await tx.user.update({
        where: { id: userId },
        data: { faceEnrolled: true, faceModel: CURRENT_FACE_MODEL, faceEnrolledAt: new Date() },
      })
    })

    return { success: true, accepted: accepted.length, total: photos.length }
  } catch (e) {
    if (e instanceof AuthServiceError) throw e
    console.error('[enroll-face]', e)
    const { code, message } = faceErrorFromException(e)
    throw new AuthServiceError(500, code, message)
  }
}

File: backend-rust/crates/streakmeet-auth/src/verification.rs
//! Email verification & password reset — parity with `backend/src/auth/verification.ts`.

use std::collections::HashMap;
use std::sync::{LazyLock, Mutex};

use chrono::{Duration, Utc};
use sqlx::PgPool;
use streakmeet_types::{ApiError, codes};

use crate::find_user_by_email;
use crate::ops::email;

static LAST_RESEND_AT: LazyLock<Mutex<HashMap<String, i64>>> =
    LazyLock::new(|| Mutex::new(HashMap::new()));

pub fn generate_token() -> String {
    use rand::RngCore;
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut bytes);
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

pub async fn issue_email_verification(
    pool: &PgPool,
    user_id: &str,
    email: &str,
) -> Result<(), ApiError> {
    let token = generate_token();
    sqlx::query(
        r#"
        UPDATE users
        SET "emailVerifyToken" = $1, "emailVerifiedAt" = NULL
        WHERE id = $2
        "#,
    )
    .bind(&token)
    .bind(user_id)
    .execute(pool)
    .await
    .map_err(|_| ApiError::new(500, codes::INTERNAL_ERROR, None))?;

    email::send_verification_email(email, &token).await
}

pub async fn mark_email_verified(pool: &PgPool, user_id: &str) -> Result<(), ApiError> {
    sqlx::query(
        r#"
        UPDATE users
        SET "emailVerifiedAt" = NOW(), "emailVerifyToken" = NULL
        WHERE id = $1
        "#,
    )
    .bind(user_id)
    .execute(pool)
    .await
    .map_err(|_| ApiError::new(500, codes::INTERNAL_ERROR, None))?;
    Ok(())
}

pub async fn verify_email_with_token(pool: &PgPool, token: &str) -> Result<(), ApiError> {
    let user_id: Option<String> =
        sqlx::query_scalar(r#"SELECT id FROM users WHERE "emailVerifyToken" = $1 LIMIT 1"#)
            .bind(token)
            .fetch_optional(pool)
            .await
            .map_err(|_| ApiError::new(500, codes::INTERNAL_ERROR, None))?;

    let Some(user_id) = user_id else {
        return Err(ApiError::new(400, codes::EMAIL_VERIFY_TOKEN_INVALID, None));
    };

    mark_email_verified(pool, &user_id).await
}

pub async fn verify_email_and_get_redirect(pool: &PgPool, token: &str) -> String {
    let app_url = std::env::var("APP_PUBLIC_URL")
        .unwrap_or_else(|_| "https://spectrmod.com".into())
        .trim_end_matches('/')
        .to_string();

    if token.is_empty() {
        return format!("{app_url}/verify-email?error=invalid");
    }

    let user_id: Option<String> =
        sqlx::query_scalar(r#"SELECT id FROM users WHERE "emailVerifyToken" = $1 LIMIT 1"#)
            .bind(token)
            .fetch_optional(pool)
            .await
            .ok()
            .flatten();

    let Some(user_id) = user_id else {
        return format!("{app_url}/verify-email?error=invalid");
    };

    let _ = mark_email_verified(pool, &user_id).await;
    format!("{app_url}/verify-email?verified=1")
}

pub async fn resend_verification(
    pool: &PgPool,
    user_id: &str,
) -> Result<serde_json::Value, ApiError> {
    let now = Utc::now().timestamp_millis();
    {
        let map = LAST_RESEND_AT.lock().unwrap();
        if let Some(last) = map.get(user_id)
            && now - last < 60_000
        {
            return Err(ApiError::new(429, codes::RESEND_COOLDOWN, None));
        }
    }

    #[derive(sqlx::FromRow)]
    struct Row {
        email: String,
        email_verified_at: Option<chrono::NaiveDateTime>,
        password_hash: String,
    }

    let user = sqlx::query_as::<_, Row>(
        r#"
        SELECT email, "emailVerifiedAt" AS email_verified_at, "passwordHash" AS password_hash
        FROM users WHERE id = $1
        "#,
    )
    .bind(user_id)
    .fetch_optional(pool)
    .await
    .map_err(|_| ApiError::new(500, codes::INTERNAL_ERROR, None))?;

    let Some(user) = user else {
        return Ok(serde_json::json!({ "success": true }));
    };

    if user.password_hash.is_empty() || user.email_verified_at.is_some() {
        return Ok(serde_json::json!({ "success": true }));
    }

    issue_email_verification(pool, user_id, &user.email).await?;
    LAST_RESEND_AT
        .lock()
        .unwrap()
        .insert(user_id.to_string(), now);

    Ok(serde_json::json!({ "success": true }))
}

pub async fn forgot_password(pool: &PgPool, email: &str) -> Result<serde_json::Value, ApiError> {
    if email.is_empty() || !email.contains('@') {
        return Err(ApiError::new(400, codes::INVALID_EMAIL, None));
    }

    let normalized = email.to_lowercase().trim().to_string();
    let user = find_user_by_email(pool, &normalized).await?;

    let Some(user) = user else {
        return Ok(serde_json::json!({ "success": true }));
    };

    if user.deleted_at.is_some() {
        return Ok(serde_json::json!({ "success": true }));
    }

    if user.password_hash.is_empty() {
        return Err(ApiError::new(400, codes::OAUTH_ACCOUNT_NO_PASSWORD, None));
    }

    let token = generate_token();
    let expires = Utc::now() + Duration::hours(1);

    sqlx::query(
        r#"
        UPDATE users
        SET "passwordResetToken" = $1, "passwordResetExpires" = $2
        WHERE id = $3
        "#,
    )
    .bind(&token)
    .bind(expires)
    .bind(&user.id)
    .execute(pool)
    .await
    .map_err(|_| ApiError::new(500, codes::INTERNAL_ERROR, None))?;

    email::send_password_reset_email(&user.email, &token).await?;
    Ok(serde_json::json!({ "success": true }))
}

pub async fn reset_password(
    pool: &PgPool,
    token: Option<&str>,
    password: Option<&str>,
) -> Result<serde_json::Value, ApiError> {
    let (token, password) = match (
        token.filter(|s| !s.is_empty()),
        password.filter(|s| !s.is_empty()),
    ) {
        (Some(t), Some(p)) => (t, p),
        _ => return Err(ApiError::new(400, codes::MISSING_FIELD, None)),
    };

    if password.len() < 8 {
        return Err(ApiError::new(400, codes::PASSWORD_TOO_SHORT, None));
    }

    let user_id: Option<String> = sqlx::query_scalar(
        r#"
        SELECT id FROM users
        WHERE "passwordResetToken" = $1 AND "passwordResetExpires" > NOW()
        LIMIT 1
        "#,
    )
    .bind(token)
    .fetch_optional(pool)
    .await
    .map_err(|_| ApiError::new(500, codes::INTERNAL_ERROR, None))?;

    let Some(user_id) = user_id else {
        return Err(ApiError::new(
            400,
            codes::PASSWORD_RESET_TOKEN_INVALID,
            None,
        ));
    };

    let password_hash =
        bcrypt::hash(password, 12).map_err(|_| ApiError::new(500, codes::INTERNAL_ERROR, None))?;

    sqlx::query(
        r#"
        UPDATE users
        SET "passwordHash" = $1, "passwordResetToken" = NULL, "passwordResetExpires" = NULL
        WHERE id = $2
        "#,
    )
    .bind(&password_hash)
    .bind(&user_id)
    .execute(pool)
    .await
    .map_err(|_| ApiError::new(500, codes::INTERNAL_ERROR, None))?;

    Ok(serde_json::json!({ "success": true }))
}

pub use crate::ops::face::EnrollFaceResult;
